RHIDP-13141: Custom review pages in Orchestrator (#2136)
redhat-developer/red-hat-developers-documentation-rhdh: modules/extend_orchestrator-in-rhdh/ref-custom-review-page-interfaces-and-utilities.adoc

:_mod-docs-content-type: REFERENCE

[id="custom-review-page-interfaces-and-utilities_{context}"]

> = Custom review page interfaces and utilities

Reviewer: ```suggestion
= Custom review page API reference
```

[role="_abstract"]

> The custom review page API includes the `ReviewComponentProps` interface, helper utilities for data processing, and UI components for implementing review pages in {product} Orchestrator workflows.

Reviewer: ```suggestion
The custom review page API provides the `ReviewComponentProps` interface, helper utilities for data processing, and UI components to implement custom review pages for {product} Orchestrator workflows.
```

== ReviewComponentProps interface

Your custom review component receives the following properties through the `ReviewComponentProps` interface:

[cols="2,2,5",options="header"]
|===
|Property |Type |Description

|`busy`
|`boolean`

> |A boolean value that indicates whether a workflow run is in progress. Disable action buttons when this value is `true` to prevent duplicate submissions.

Reviewer: ```suggestion
|Indicates whether a workflow run is in progress. Disable action buttons when this value is `true` to prevent duplicate submissions.
```

|`schema`
|`JSONSchema7`

> |The JSON Schema definition for the workflow form. Use it for field structure, titles, and UI hints such as hidden fields.

Reviewer: ```suggestion
|The JSON Schema definition for the workflow form. Defines field structure, titles, and UI hints such as hidden fields.
```

> |The JSON Schema definition for the workflow form. Use this to determine field types, validation rules, and UI hints such as hidden fields.

Reviewer 2: ```suggestion
|The JSON Schema definition for the workflow form. Use it for field structure, titles, and UI hints such as hidden fields.

```

|`data`
|`JsonObject`

> |The form data submitted by the user, awaiting review before the workflow runs. This object has all form values structured according to the schema.

Reviewer: ```suggestion
|The form data submitted by the user, awaiting review before the workflow runs. Contains form values structured according to the schema.
```

|`handleBack`
|`() => void`

> |Callback to return to the previous wizard step (same behavior as the default review page). Call it when the user chooses *Back*.

Reviewer: ```suggestion
|Returns to the previous step (same behavior as the default review page). This callback matches the default review page behavior.
```

> |A callback function to navigate to the *workflow form*. Call this function when the customer clicks *Back* or *Cancel*.

Reviewer 2: ```suggestion
|Callback to return to the previous wizard step (same behavior as the default review page). Call it when the user chooses *Back*.
```

|`handleExecute`
|`() => void`

> |A callback function to run the workflow with the reviewed data. Call this function when the user clicks *Run* to start the workflow.

Reviewer: ```suggestion
|Runs the workflow with the reviewed data. Call this function when the user clicks *Run* to start the workflow.
```

> |A callback function to trigger workflow execution with the reviewed data. Call this function when the user confirms the data and clicks *Enter*.

Reviewer 2: ```suggestion
|A callback function to trigger workflow execution with the reviewed data. Call this function when the user clicks Run to start workflow execution.
```

|===

== Helper utilities

The `orchestrator-form-react` package exports the following utilities to help you build custom review pages that handle data correctly:

[cols="2,3,4",options="header"]
|===
|Function |Signature |Description

|`generateReviewTableData`
|`(schema: JSONSchema7, data: JsonObject, options?: {includeHiddenFields?: boolean}) => JsonObject`

> |Processes form data for display by respecting `ui:hidden` fields, masking password fields, and structuring nested data. Use this to prepare data for rendering in your custom review component.

Reviewer: ```suggestion
|Processes form data for display. Respects `ui:hidden` fields, masking password fields, and structures nested data. Use this to prepare data for rendering in your custom review component.
```

|`schemaHasUiHiddenFields`
|`(schema: JSONSchema7) => boolean`

> |Returns `true` if the schema contains fields marked with `ui:hidden` in the UI schema. Use this to determine whether to display a toggle for showing hidden fields.

Reviewer: ```suggestion
|Returns `true` if the schema contains fields marked with `ui:hidden` in the UI schema. Use this to determine if the UI should display a toggle for hidden fields.
```
